### ISSUE TITLE
[Bug] Whisper transcription truncated after ~16 tokens (any clip with >~4s of continuous speech)

### ISSUE THREAD

**Issue by Anjielon** [FastFlowLM/FastFlowLM#698]
# [Bug] Whisper transcription truncated after ~16 tokens (any clip > ~4s of continuous speech)

## Summary
On Linux (Strix Halo, XDNA2/NPU2), `/v1/audio/transcriptions` returns only the
first ~15-16 tokens of any transcription. Reproduced on the released 1.0.3
binary AND on a from-source build (same commit as `main`).

## Reproduction (deterministic)
Italian TTS clip, 7.3 s, 20 words, 16 kHz mono. Three different clip lengths
(3.2 s / 5 s / 7.3 s) all return the SAME 50-character prefix; changing
`response_format` (json / verbose_json / text / srt), `language`, or
`--ctx-len` has no effect:

    " Domani alle 9 accendi le luci del soggiorno e ric"   ← always 50 chars

## Root cause
`src/common/whisper/modeling_whisper.cpp`, decode loop of `generate()` (and
`generate_streaming()`):

    int watching_dog = 16;
    ...
    if (watching_dog == 0 && allow_force_time_stamp){
        last_idx = this->_sample_in_time_stamp(logits);   // forces a timestamp
        ...

After 16 content tokens without a timestamp, the loop force-samples a timestamp
token. Whisper then treats the segment as closed and emits EOT — so any
continuous-speech segment longer than ~16 tokens (~4 s of speech) is cut.

## Fix (verified on device)
Raising the watchdog to the decode maximum removes the truncation with no other
change (both occurrences):

    int watching_dog = 448;

Same 7.3 s clip after the fix: 115 chars, full sentence, 5.0 s on NPU2
(Ryzen AI MAX+ 395, FW 1.1.2.65, amdxdna 0.6, kernel 7.0.8):

    " Domani alle 9 accendi le luci del soggiorno e ricordami di comprare
      il latte e le uova al supermercato e se lunga."

If the watchdog exists to protect against runaway no-timestamp decoding, a
less invasive form would be to only force a timestamp when the model has
produced no *content* tokens either, or to make the threshold configurable.
Happy to open a PR with either form.

## Environment
- FastFlowLM 1.0.3 (release tarball) and from-source (main)
- Fedora 43, kernel 7.0.8, amdxdna 0.6.0, NPU FW 1.1.2.65
- Ryzen AI MAX+ 395 (Strix Halo), memlock unlimited

**Comment by huppiflupp** [comment]
Confirming this on a second Strix Halo box, and adding numbers for German and Mandarin: the truncation costs more accuracy than the NPU path gains.

Setup: Ryzen AI Max+ 395 (gfx1151, XDNA2), Nobara 44 / kernel 7.2, in-tree `amdxdna`, XRT 2.26, FLM v1.0.1 release binary, `flm serve lfm2.5-it:1.2b --asr 1`, `whisper-v3:turbo`, `/v1/audio/transcriptions`, no `language` set. 60 FLEURS dev clips (30 `de_de`, 30 `cmn_hans_cn`, 12 minutes total), reference text from the dataset. Baseline: the same model as `faster-whisper` `large-v3-turbo`, int8 on the CPU of the same machine.

| | German WER | Mandarin CER | speed |
|---|---|---|---|
| CPU, faster-whisper large-v3-turbo | **4.1 %** | **10.5 %** | 3.0x realtime |
| NPU, FLM whisper-v3:turbo | 23.0 % | 16.5 % | 3.8x / 2.7x realtime |

12 of 60 transcripts are cut, and where they are not cut the NPU output matches the CPU output closely, so the error is the truncation, not the recognition.

Two observations that may narrow the trigger:

- It is not simply "clips longer than ~4 s". Of the 34 clips longer than 10 s, only 5 are cut, while several 4–8 s clips lose most of their text. Examples (German, 8.1 s): `Einige Patienten haben sich möglicherweise im Krankenhaus mit dem Erreger ang` — the CPU returns the full sentence, 164 characters. Mandarin, 8.0 s: `咸水蛾的主要棲息地并非海` (12 of 26 characters). The cut-off transcripts end mid-word, with no punctuation.
- The cut length varies (12 to ~120 characters), which fits a watchdog that counts decoded tokens rather than audio seconds: a clip whose speech starts with a pause or an audible breath reaches 16 content tokens later in the audio.

Happy to re-run the 60-clip set against a build with `watching_dog = 448` if that helps confirm the fix; the harness and the FLEURS sample list are small and I can share them.
